### PR TITLE
test: Pagination component

### PR DESCRIPTION
**Pull request by yardz**
### SUMMARY
- Divided the components into different files.
- Created the tests for the Pagination components

### TEST PLAN
No change in behavior occurred. So all existing tests must pass normally.

### CODE REVIEW

File: superset-frontend/src/components/Pagination/Ellipsis.test.tsx
Comment (geido) [inline superset-frontend/src/components/Pagination/Ellipsis.test.tsx:29]: We are trying to use the user-event from @testing-library/user-event wherever it is possible. I think we can use it here as well.

File: superset-frontend/src/components/Pagination/Item.test.tsx
Comment (geido) [inline superset-frontend/src/components/Pagination/Item.test.tsx:33]: Same as my other comment concerning user-event

**Review by geido — CHANGES_REQUESTED** [review]
Hey @yardz this looks very nice. My only comment is concerning the usage of user-event. I commented two below but that same comment applies everywhere. Thank you!

File: superset-frontend/src/components/Pagination/Ellipsis.test.tsx
Comment (yardz) [reply superset-frontend/src/components/Pagination/Ellipsis.test.tsx]: Thanks for the tip, I didn't know `user-event` but I liked it.

File: superset-frontend/src/components/Pagination/Item.test.tsx
Comment (yardz) [reply superset-frontend/src/components/Pagination/Item.test.tsx]: Done!

**Comment by yardz** [comment]
@geido When you can approve, I appreciate :-D

**Comment by yardz** [comment]
@nytai can you look at this too?
rsrsrsrs
Thank you for the support!!! :-D

**Comment by junlincc** [comment]
@yardz Hey Bruno, sure everyone can review and comment, but Evan and Jesse are the best persons to approve your PRs. It will be great to have blessings from both of them. Hope you don't mind iterating your code and solution based on feedback from the open, to me, this really is the best part being part of an OSS project. 🙏

File: superset-frontend/src/components/Pagination/PaginationButton.types.ts
Comment (ktmud) [inline superset-frontend/src/components/Pagination/PaginationButton.types.ts:20]: I think you can just call this file `./components/Pagination/types.ts`. I haven't seen this filename pattern (`ComponentName.types.ts`) in Superset.

Maybe also call the type `PaginationButtonProps` to avoid being confused with it being an actual React component.

File: superset-frontend/src/components/Pagination/PaginationWrapper.test.tsx
Comment (ktmud) [inline superset-frontend/src/components/Pagination/PaginationWrapper.test.tsx:28]: `ThemProvider` and `import '@testing-library/jest-dom'` would not be needed if you just `import { render } from 'spec/helpers/testing-library'`.

File: superset-frontend/src/components/Pagination/index.ts
Comment (ktmud) [inline superset-frontend/src/components/Pagination/index.ts:24]: I think you can just `export default Pagination` here and do assign `Next, Prev, Item, Ellipsis` to the functional component, as the original component name `Pagination` is more natural than `PaginationWrapper`.

```ts
type PaginationChild = Next | Prev | Item | Ellipsis;

function Pagination({
  children
} : {
  children: PaginationChild | PaginationChild[]
}) {
  return <PaginationList role="navigation">{children}</PaginationList>;
}

Pagination.Next = Next;
Pagination.Prev = Prev;
Pagination.Item = Item;
Pagination.Ellipsis = Ellipsis;

export default Pagination;
```

**Review by ktmud — COMMENTED** [review]
In general I'm in favor of splitting files, but this component seems like a perfect case where multiple components in one file could be allowed.

File: superset-frontend/src/components/Pagination/PaginationButton.types.ts
Comment (yardz) [reply superset-frontend/src/components/Pagination/PaginationButton.types.ts:20]: I usually do this when it's just "Types". If you want to rename `PaginationButtonProps .ts` it is better.

File: superset-frontend/src/components/Pagination/PaginationWrapper.test.tsx
Comment (yardz) [reply superset-frontend/src/components/Pagination/PaginationWrapper.test.tsx:28]: They are necessary...

No ThemeProvider:
Captura de Tela 2021-02-23 às 11 01 45 AM

No `@testing-library/jest-dom`:
Captura de Tela 2021-02-23 às 11 02 31 AM

Comment (yardz) [reply superset-frontend/src/components/Pagination/PaginationWrapper.test.tsx:28]: Very good to know :-)

File: superset-frontend/src/components/Pagination/index.ts
Comment (yardz) [reply superset-frontend/src/components/Pagination/index.ts:24]: I chose this way because it is easier to understand the code. Many develops have difficulty understanding different types for `children`.
And doing assignments like that is not the kind of thing that someone catches his eye and understands. (And it is more difficult to get to the "original component").
To tell you the truth, I only see disadvantages in using the way you proposed...

Comment (yardz) [reply superset-frontend/src/components/Pagination/index.ts:24]: @ktmud Do you agree or still want me to make the changes?

Comment (ktmud) [reply superset-frontend/src/components/Pagination/index.ts:24]: Adding typing to `children` was just another suggestion unrelated to code organization. Having custom types for `children` enforces the usage of `<Pagination>...</Pagination>` so that TypeScript can warn you when other children types are used. E.g.  this code

```tsx
<Pagination>
  <li className="pagi">Select control</li>
</Pagination>
```

should show an error because an arbitrary `childNode` will either render incorrectly or create customization outside of the component itself, defeating the purpose of component abstraction.

Re: `PaginationWrapper` vs `Pagination`, it really doesn't matter that much. But when the component is used outside of `createUltimatePagination`, do you really think

```ts
import Pagination from 'src/components/pagination';

<Pagination.PaginationWrapper>
  <Pagination.Prev />
  <Pagination.Prev />
</Pagination.PaginationWrapper>
```

reads better than

```ts
import Pagination from 'src/components/pagination';

<Pagination>
  <Pagination.Prev />
  <Pagination.Prev />
</Pagination>
```

?

I don't get your point about assignments.

I'll defer to @nytai on which style to use since he's the original author of this component. TBH the original class component style also works perfectly fine for me.

As a general note, if all we wanted was adding tests to make sure components work correctly, I'd refrain from doing pure code style refactoring unless there are clear benefits to do so.

File: superset-frontend/src/components/Pagination/PaginationButton.types.ts
Comment (ktmud) [reply superset-frontend/src/components/Pagination/PaginationButton.types.ts]: Creating a separate file for a single React props interface is not a pattern we've seen in Superset, and there are many more places that uses `types.ts` or a `types/` folder with larger files encompassing a set of related types.

I'd recommend follow the convention.

File: superset-frontend/src/components/Pagination/index.ts
Comment (yardz) [reply superset-frontend/src/components/Pagination/index.ts:24]: > should show an error because an arbitrary `childNode` will either render incorrectly or create customization outside of the component itself, defeating the purpose of component abstraction.

I agree, that really makes sense.

Comment (yardz) [reply superset-frontend/src/components/Pagination/index.ts:24]: > Re: `PaginationWrapper` vs `Pagination`, it really doesn't matter that much. But when the component is used outside of `createUltimatePagination`, do you really think

The second example is really better but at the same time I am bothered to put something like "Pagination.Next = Next" because in my opinion this would need to be tested as well (which is an option).
This code also doesn't seem like something that a person with little experience will understand, and it is also more "difficult" to find the original component. I understand that there are benefits but the bad part seems to me to be greater than the benefit.

Perhaps changing `PaginationWrapper` to `Container` will improve that part, like this:
```ts
import Pagination from 'src/components/pagination';

<Pagination.Container>
  <Pagination.Prev />
  <Pagination.Prev />
</Pagination.Container>
```
It's not that good but it's close...

Comment (yardz) [reply superset-frontend/src/components/Pagination/index.ts:24]: > I'll defer to @nytai on which style to use since he's the original author of this component. TBH the original class component style also works perfectly fine to me.

I agree, it makes sense to me.

Comment (yardz) [reply superset-frontend/src/components/Pagination/index.ts:24]: > As a general note, if all we wanted was adding tests to make sure components work correctly, I'd refrain from doing pure code style refactoring unless there are clear benefits to do so.

I agree that it works. However, changing classes to functions improves reading, facilitates maintenance, facilitates testing, code minification (and in some cases performance too). I personally avoid class components as much as possible so I think that changing from class to function is always an improvement.
But I understand your point

Comment (ktmud) [reply superset-frontend/src/components/Pagination/index.ts:24]: > Perhaps changing PaginationWrapper to Container will improve that part, like this:

We see fairly often the `<Component><Component.Child /></Component>` pattern in Superset (probably a byproduct of AntD), e.g. `Menu` and `Menu.Item`, `Panel` and `Panel.Heading`, `Panel.Body`, so I really don't see the need of breaking the convention here. I'm not sure why it is difficult to understand what the assignment statement does. Extra test is not needed if you just use `Pagination.Next` etc in the tests as well.

If you think the code is hard to understand to new comers, maybe add a comment like this:

```ts
// Expose child components as component attribute to enforce namespacing
// and avoid the need to import them separately.
Pagination.Next  = Next;
Pagination.Prev  = Prev;
```

Comment (nytai) [reply superset-frontend/src/components/Pagination/index.ts:24]: I have no strong opinions here. Some history: this was implemented as a quickfix once the previous component broke after we upgraded react-bootstrap. The api was designed to mirror the newer react bootstrap pagination api https://react-bootstrap.github.io/components/pagination/ and it's also in line with the compound component pattern heavily used by AntD.

Comment (yardz) [reply superset-frontend/src/components/Pagination/index.ts:24]: @ktmud 
> ```ts
> type PaginationChild = Next | Prev | Item | Ellipsis;
> ```

I tried to do this but it gives this error.
Captura de Tela 2021-02-24 às 9 12 36 AM

Putting the typeof looks like this:
Captura de Tela 2021-02-24 às 9 15 08 AM

I believe it is because we are working with functions here...For things like these, I think that typing in children usually gets in the way of helping. Do you think it would be interesting to keep the type?

Comment (yardz) [reply superset-frontend/src/components/Pagination/index.ts]: > ```ts
> // Expose child components as component attribute to enforce namespacing
> // and avoid the need to import them separately.
> Pagination.Next  = Next;
> Pagination.Prev  = Prev;
> ```

@ktmud Still, done

Comment (ktmud) [reply superset-frontend/src/components/Pagination/index.ts]: This should work:

```ts
import { ReactElement } from 'react';

type PaginationChild =
  | ReactElement<NextProps, typeof Next>
  | ReactElement<PrevProps, typeof Props>
```

But feel free to skip this change it it's too much work.

Comment (yardz) [reply superset-frontend/src/components/Pagination/index.ts]: > ```ts
> import { ReactElement } from 'react';
> 
> type PaginationChild =
>   | ReactElement<NextProps, typeof Next>
>   | ReactElement<PrevProps, typeof Props>
> ```

I tried this but it really didn't look very readable...
I ended up doing it like this:
```ts
interface PaginationProps {
  children: JSX.Element | JSX.Element[];
}
```
It is certainly not the best way possible, but it is easy to understand.

Comment (ktmud) [reply superset-frontend/src/components/Pagination/index.ts]: Yeah, but the only thing this prevents is empty or text child nodes as it doesn't check props. Since this component normally isn't directly used, I don't think this matters that much so I'm OK with either way.

Just an idea for thought on how you can enforce children type in the future.

**Review by ktmud — COMMENTED** [review]
This looks good. Thanks for the iterations @yardz !

I'll leave it to @nytai to stamp and merge as he's the original owner of this code.

File: superset-frontend/src/components/Pagination/index.tsx
Comment (rusackas) [inline superset-frontend/src/components/Pagination/index.tsx:31]: I see a bunch of pixel values in the styles here that should be changed to `gridUnit` but have no business holding up this PR, so I'll apply the `revisit: theme vars` label here to tackle later as a rainy day or onboarding task.

**Review by rusackas — APPROVED** [review]
Well, I'm _super_ late to the party on this thread, sorry about that. Looks pretty great from here though. I'm on the fence about exactly how granular we should break up our component files, but this looks perfectly clean/readable/tested to me. Thanks to all involved for the input and hard work on this.

**Pull request merged**
